### PR TITLE
fix: Allow variable size in ONNX inputs

### PR DESCRIPTION
**Pull request by memoryz**
## Related Issues/PRs

This should fix issue #1848.

## What changes are proposed in this pull request?

This PR adds support to infer the shape from the batched value passed into the ONNXModel. Two constraints apply:
1. At most 1 dimension can be variable size (-1) specified in the input node's tensor info.
2. If batch size is fixed and greater than 1, the values provided in the batch must all have the same shapes, otherwise we cannot infer the proper tensor shape.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change any dependencies?

- [x] No. You can skip this section.
- [ ] Yes. Make sure the dependencies are resolved correctly, and list changes here.

## Does this PR add a new feature? If so, have you added samples on website?

- [x] No. You can skip this section.
- [ ] Yes. Make sure you have added samples following below steps.

1. Find the corresponding markdown file for your new feature in `website/docs/documentation` folder.
   Make sure you choose the correct class `estimators/transformers` and namespace.
3. Follow the pattern in markdown file and add another section for your new API, including pyspark, scala (and .NET potentially) samples.
4. Make sure the `DocTable` points to correct API link.
5. Navigate to website folder, and run `yarn run start` to make sure the website renders correctly.
6. Don't forget to add `<!--pytest-codeblocks:cont-->` before each python code blocks to enable auto-tests for python samples.
7. Make sure the `WebsiteSamplesTests` job pass in the pipeline.

### CODE REVIEW

**Comment by memoryz** [comment]
Please don't merge this PR. More changes needed.

File: deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXModelSuite.scala [hunk 1]
```diff
@@ -396,4 +396,43 @@ class ONNXModelSuite extends TestBase
...
+  private lazy val testDfText: DataFrame = {
+    Seq(
+      Tuple1(Array("A", "B", "C")),
+      Tuple1(Array("A", "B", "B", "C", "A"))
+    ).toDF("features")
+  }
+
+  test("ONNX model can accept variable size input if batch size is set to 1"){
+    val Array(row1, row2) = testTextModel
+      .setMiniBatchSize(1) // If the array size in each row can vary, must set batch size to 1.
```
Comment (mhamilton723) [inline deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXModelSuite.scala:418]: nit: just a friendly reminder to add this suggestion to the appropriate error message so people know that it is supported if theres no auto-batching :)

**Review by svotaw — APPROVED** [review]

File: deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXModelSuite.scala (hunk 1 above)
Comment (memoryz) [reply deep-learning/src/test/scala/com/microsoft/azure/synapse/ml/onnx/ONNXModelSuite.scala]: done.

**Comment by memoryz** [comment]
@mhamilton723 ready for merge.

**Review by mhamilton723 — APPROVED** [review]
Lovely work Jason!

**Pull request merged**